onnxscript/tools/memory_peak.py

```diff
@@ -0,0 +1,243 @@
+                g.update(r)
+        if conn.poll(timeout=timeout):
+            code = conn.recv()
+            if code == -3:
+                break
+
+    # final iteration
+    end = process.memory_info().rss
+    cpu.update(end)
+    if cuda:
+        for r, g in zip(gpu_used(), gpus):
+            g.update(r)
+
+    # send
+    cpu.send(conn)
+    conn.send(len(gpus))
+    for g in gpus:
+        g.send(conn)
+    if cuda:
+        nvmlShutdown()
+    conn.close()
+
+
+class MemorySpy:
+    """
+    Information about the spy. It class method `start`.
+    Method `stop` can be called to end the measure.
+
+    Args:
+        pid: process id  of the process to spy on
+        delay: spy on every delay seconds
+        cuda: enable cuda monitoring
+    """
+
+    def __init__(self, pid: int, delay: float = 0.01, cuda: bool = False):
+        self.pid = pid
+        self.delay = delay
+        self.cuda = cuda
+        self.start()
+
+    def start(self) -> MemorySpy:
+        """Starts another process and tells it to spy."""
+        self.parent_conn, self.child_conn = multiprocessing.Pipe()
+        self.child_process = multiprocessing.Process(
+            target=_process_memory_spy, args=(self.child_conn,)
+        )
+        self.child_process.start()
+        data = self.parent_conn.recv()
+        if data != -2:
+            raise RuntimeError(f"The child processing is supposed to send -2 not {data}.")
+        self.parent_conn.send(self.pid)
+        self.parent_conn.send(self.delay)
+        self.parent_conn.send(1 if self.cuda else 0)
+        data = self.parent_conn.recv()
+        if data != -2:
+            raise RuntimeError(
+                f"The child processing is supposed to send -2 again not {data}."
+            )
+        return self
+
+    def stop(self) -> dict[str, list[Monitor]]:
+        """Stops spying on."""
+        self.parent_conn.send(-3)
+
+        cpu = [Monitor.recv(self.parent_conn)]
+
+        n_gpus = self.parent_conn.recv()
+        gpus = []
+        for _ in range(n_gpus):
+            gpus.append(Monitor.recv(self.parent_conn))
+
+        self.parent_conn.close()
+        self.child_process.join()
+        res = dict(cpu=cpu)
+        if self.cuda:
+            res["gpus"] = gpus
+        return res
+
+
+def start_spying_on(
+    pid: int | None = None, delay: float = 0.01, cuda: bool = False
+) -> MemorySpy:
+    """Starts the memory spy. The function starts another
+    process spying on the one sent as an argument.
+
+    Args:
+        pid: process id to spy or the the current one.
+        delay: delay between two measures.
+        cuda: True or False to get memory for cuda devices
+
+    Example::
+
+    .. code-block:: python
+
+        from onnxscript.tools.memory_peak import start_spying_on, flatten
+
+        p = start_spying_on()
+        # ...
+        # code to measure
+        # ...
+        stat = p.stop()
+        print(stat)
+        print(flatten(stat))
```

Comment:
nit: Example should come before the Args section (https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)

Reply:
done